// electron-builder/out/util.js
"use strict";

const child_process_1 = require("child_process");
const bluebird_1 = require("bluebird");
const readPackageJsonAsync = require("read-package-json");
const os = require("os");
const path = require("path");
exports.log = console.log;
exports.DEFAULT_APP_DIR_NAME = "app";
exports.commonArgs = [{
    name: "appDir",
    type: String,
    description: "Relative (to the working directory) path to the folder containing the application package.json. Working directory or app/ by default."
}];
exports.readPackageJson = bluebird_1.Promise.promisify(readPackageJsonAsync);
function installDependencies(appDir, electronVersion) {
    let arch = arguments.length <= 2 || arguments[2] === undefined ? process.arch : arguments[2];
    let command = arguments.length <= 3 || arguments[3] === undefined ? "install" : arguments[3];

    exports.log("Installing app dependencies for arch %s to %s", arch, appDir);
    const gypHome = path.join(os.homedir(), ".electron-gyp");
    const env = Object.assign({}, process.env, {
        npm_config_disturl: "https://atom.io/download/atom-shell",
        npm_config_target: electronVersion,
        npm_config_runtime: "electron",
        npm_config_arch: arch,
        HOME: gypHome,
        USERPROFILE: gypHome
    });
    let npmExecPath = process.env.npm_execpath || process.env.NPM_CLI_JS;
    const npmExecArgs = [command, "--production"];
    if (npmExecPath == null) {
        npmExecPath = "npm";
    } else {
        npmExecArgs.unshift(npmExecPath);
        npmExecPath = process.env.npm_node_execpath || process.env.NODE_EXE || "node";
    }
    return spawn(npmExecPath, npmExecArgs, {
        cwd: appDir,
        stdio: "inherit",
        env: env
    });
}
exports.installDependencies = installDependencies;
function exec(file, args, options) {
    return new bluebird_1.Promise((resolve, reject) => {
        child_process_1.execFile(file, args, options, function (error, stdout, stderr) {
            if (error == null) {
                resolve([stdout, stderr]);
            } else {
                if (stdout.length !== 0) {
                    console.error(stdout.toString());
                }
                if (stderr.length !== 0) {
                    console.error(stderr.toString());
                }
                reject(error);
            }
        });
    });
}
exports.exec = exec;
function spawn(command, args, options) {
    return new bluebird_1.Promise((resolve, reject) => {
        const p = child_process_1.spawn(command, args, options);
        p.on("close", code => code === 0 ? resolve() : reject(new Error(command + " exited with code " + code)));
    });
}
exports.spawn = spawn;
function getElectronVersion(packageData, filePath) {
    const devDependencies = packageData.devDependencies;
    let electronPrebuiltDep = devDependencies == null ? null : devDependencies["electron-prebuilt"];
    if (electronPrebuiltDep == null) {
        const dependencies = packageData.dependencies;
        electronPrebuiltDep = dependencies == null ? null : dependencies["electron-prebuilt"];
    }
    if (electronPrebuiltDep == null) {
        throw new Error("Cannot find electron-prebuilt dependency to get electron version in the '" + filePath + "'");
    }
    return electronPrebuiltDep.substring(1);
}
exports.getElectronVersion = getElectronVersion;
